rfcs/0005-version-matrix.md

```diff
@@ -0,0 +1,477 @@
+When they submit a simple PR we don't want them being hit with unrelated error messages in the CI.
+Such a smooth experience will no doubt also be appreciated by veteran contributors.
+Of the numerous times we've been hit by this, take for example the case where Raph had to start [debugging an unrelated CI error](https://github.com/linebender/vello/pull/401/commits/9861e506dc89924cb6feb30610264f7266e467e3) during his numerical robustness work.
+These scenarios ramp up in frequency very fast in our low traffic projects like Druid or Piet.
+In cases where a [simple one line fix PR](https://github.com/linebender/piet/pull/561) could be accepted we're instead greeted with the task of having to deal with the accumulated decay of the last 6 months.
+
+In similar vein it would be great if contributors could run various tests locally in a manner that approximates the CI, to cut down on surprising errors later.
+
+### Historic build reproduction
+
+Sometimes you want to check out an older revision of a project and compile it.
+For example if there is a newly found bug, but it was introduced a while ago and you want to debug it in the version with the least changes.
+It would be very useful indeed to be able to compile this historic revision without having to start dealing with the fact that some unknown sub-dependency's newer version isn't compatible and causes unexpected further issues.
+
+### We want to support more than just the latest stable Rust toolchain
+
+Not everyone can or will upgrade their Rust toolchain the moment a new version is published.
+This includes potential new contributors, to whom we want to let write their patch without having to start updating their toolchain.
+Beyond that, it will be even more common among application developers who use our libraries.
+It's a question of balance of course, but we want to support toolchains as old as possible.
+
+### We want a widely applicable strategy
+
+In the absence of a well thought out general strategy we will have (potentially ephemeral) ad hoc approaches for every distinct project.
+This means an increased maintenance burden, as problems solved with one approach won't necessarily carry over to another.
+Also there will be more problems to solve, because a mix of strategies will bring us the union of their respective problems.
+
+
+## User-facing explanation
+
+We test and support only the latest versions of all our package dependencies.
+
+Our minimum supported Rust version (MSRV) is largely decided by our dependencies and we just document the fact.
+
+
+## Implementation strategy
+
+### Keep the existing general Linebender CI as a base and extend it
+
+Our current general CI script is the result of years of experience.
+It already has good techniques to avoid some sources of dependency contamination, e.g. feature activations caused by dev dependencies.
+It also already pins the stable Rust toolchain version, allowing us to choose when to update it.
```

Comment:
IMO this should be unpinned. Most people *do* update almost as soon as a new rust version is release (< 1 week, same day is common), and it's important that our libs work with the new version as soon as possible. If clippy is causing issues then the version could be pinned specifically for the clippy check.

Reply:
There is nothing in this strategy that is stopping us from updating the pin the second a new toolchain is released.

Unpinning it would just mean that the errors interfere with unrelated PRs. Someone submiting a typo fix to the docs shouldn't have to start dealing with new toolchain issues.

We ran unpinned toolchains for 4 years in the Linebender repos. Plenty of experience with that. A year ago there was [a proposal to start pinning](https://github.com/linebender/glazier/pull/76) them. I personally was hesitant to do it at first, but after discussing it in more depth in that issue and at office hours we decided to give it a try. Now it's one year later and I can say that it has been a big success. We have had zero cases where the new toolchain failed to compile our code (which might stay hidden until the pin is updated) but we've managed to avoid unrelated warnings and errors adding friction to PRs.

Reply:
I will also add that our standard CI already has a docs job with an unpinned nightly toolchain. It won't catch everything, but it does work as a partial canary for certain issues coming from new Rust toolchain features.